### PR TITLE
ReachabilityRule as Claim

### PR DESCRIPTION
**Pull request by ana-pantilie**
Checkpoint in https://github.com/kframework/kore/issues/906

- `kore-exec` and `kore-repl` now use `ReachabilityRule Variable`
- Replaces the `Coercible` constaints in `Claim` with `ToRulePattern` and `FromRulePattern` on `goal` and `Rule goal`
- Adds `goalToRule` and `ruleToGoal` as methods in `Goal`.
- `Rule (ReachabilityRule Variable)` is now a newtype over `RewriteRule Variable`.
- `Goal.strategy` now returns a `Stream` to make a general version of `verifyClaimStep` typesafe when extracting the first and followup steps from the `Strategy` list/stream. This needed a new package dependency, will revert if you think it's a bit overkill.
- Added `Test.Kore.Strategies.Reachability.Verification`.

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/src/Kore/Exec.hs [hunk 1]
```diff
@@ -526,25 +521,26 @@ assertSomeClaims claims =
         ++  "on the representation of claims."
 
 makeClaim
-    :: Coercible (RulePattern Variable) claim
+    :: Goal.FromRulePattern claim
+    => Goal.ToRulePattern claim
     => (Attribute.Axiom, claim) -> claim
 makeClaim (attributes, rule) =
```
Comment (ttuegel) [inline kore/src/Kore/Exec.hs]:
```suggestion
makeClaim (attributes, Goal.toRulePattern -> rule) =
```

File: kore/src/Kore/Exec.hs [hunk 2]
```diff
@@ -648,13 +644,16 @@ initializeProver definitionModule specModule within =
...
         simplifiedSpecClaims <-
             mapM (mapMSecond simplifyToList) specClaims
         specAxioms <- Profiler.initialization "simplifyRuleOnSecond"
             $ traverse simplifyRuleOnSecond (concat simplifiedSpecClaims)
-        let
-            axioms = coerce <$> rewriteRules
-            claims = fmap makeClaim specAxioms
+        let claims = fmap makeClaim specAxioms
+            axioms =
+                Goal.fromRulePattern (Goal.goalToRule firstClaim)
```
Comment (ttuegel) [inline kore/src/Kore/Exec.hs]: Why are we inserting the first claim into the list of axioms?

File: kore/src/Kore/Repl.hs [hunk 3]
```diff
@@ -193,10 +190,12 @@ runRepl axioms' claims' logger replScript replMode outputFile = do
...
-    addIndexesToClaims len cls =
-        fmap
-            (coerce . addIndex)
-            (zip (fmap coerce cls) [len..])
+    addIndexesToClaims len claims'' =
+        let toAxiomAndBack (claim', index) =
+                ruleToGoal
+                    claim'
+                    $ addIndex (goalToRule claim', index)
+        in fmap toAxiomAndBack (zip claims'' [len..])
```
Comment (ttuegel) [inline kore/src/Kore/Repl.hs]:
```suggestion
        in zipWith toAxiomAndBack claims'' [len..]
```

File: kore/src/Kore/Step/Rule.hs [hunk 4]
```diff
@@ -916,3 +922,45 @@ applySubstitution substitution rule =
...
+
+instance ToRulePattern (ImplicationRule Variable)
+
+instance ToRulePattern (ReachabilityRule Variable) where
+    toRulePattern (OnePath rule) = toRulePattern rule
+    toRulePattern (AllPath rule) = toRulePattern rule
+
+-- | We need to know the context when transforming a 'RulePattern Variable'
+-- into a 'rule', hence the first 'rule' argument. In general it can be ignored
+-- when the 'rule' and the 'RulePattern Variable' are representationally equal.
```
Comment (ttuegel) [inline kore/src/Kore/Step/Rule.hs]: Can the first argument be a `Proxy` then?

File: kore/src/Kore/Strategies/Goal.hs [hunk 5]
```diff
@@ -164,7 +188,7 @@ class Goal goal where
         :: goal
         -> [goal]
         -> [Rule goal]
-        -> [Strategy (Prim goal)]
+        -> Stream (Strategy (Prim goal))
```
Comment (ttuegel) [inline kore/src/Kore/Strategies/Goal.hs]: Why this change?

File: kore/app/exec/Main.hs [hunk 6]
```diff
@@ -449,7 +445,7 @@ koreProve execOptions proveOptions = do
                     Bounded.Failed final -> return (failure final)
             else
                 either failure (const success)
-                <$> prove stepLimit mainModule specModule (Proxy @(OnePathRule Variable))
+                <$> prove stepLimit mainModule specModule (Proxy @(ReachabilityRule Variable))
```
[the remark this answers is not in the document]
Comment (ana-pantilie) [reply kore/app/exec/Main.hs]: Doing this will mostly revert https://github.com/kframework/kore/pull/1162, just want to make sure that's what we want.

File: kore/src/Kore/Exec.hs (hunk 2 above)
Comment (ana-pantilie) [reply kore/src/Kore/Exec.hs]: This doesn't insert the claim, we just need to get the axiom type here which is determined by the claim type. If we revert the generalization of `Exec.prove` I think this awkward way of getting the axiom type at start-up won't be needed anymore.

File: kore/src/Kore/Strategies/Goal.hs (hunk 5 above)
Comment (ana-pantilie) [reply kore/src/Kore/Strategies/Goal.hs]: This was for generalizing `verifyClaimStep`. I think there were 2 ways to solve this:
  1. assume `strategy` returns an infinite list and then `verifyClaimStep` can extract steps from it
  2.  add `strategyFirstStep` and `strategySecondStep` to the class but that would assume all claim types have at maximum 2 different steps (but it's very possible that we wouldn't want something more general anyway)

I opted for 1 since it seemed more general and easier to refactor. I can do 2 if that makes more sense, or maybe there is a better way.

File: kore/app/exec/Main.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ttuegel) [reply kore/app/exec/Main.hs]: I think that's OK, the goal of #1162 was to facilitate this change now. I don't think we can generalize `prove` beyond reachability logic, and that has only `OnePath` and `AllPath` claims.

File: kore/src/Kore/Strategies/Goal.hs (hunk 5 above)
Comment (ttuegel) [reply kore/src/Kore/Strategies/Goal.hs]: I agree, returning an infinite stream as the strategy makes sense!

File: kore/src/Kore/Step/Rule.hs (hunk 4 above)
Comment (ana-pantilie) [reply kore/src/Kore/Step/Rule.hs]: I don't think it can, or I just don't know how to do this with a `Proxy`. I need a case on that first argument, if it's `OnePathRule` apply the `OnePath` constructor and if it's an `AllPathRule` apply the `AllPath` constructor.
